Fixed #12978 -- Added support for RSS feed stylesheets.

bmispelon opened the pull request:
# Trac ticket number
ticket-35420 (duplicate)
ticket-12978 (the actual original ticket)

# Branch description
To simplify the implementation I decided to restrict the types of stylesheets to XSL (it lets me hardcode the type of the stylesheet), it seems to be the de-facto standard anyway.

# Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

felixxm commented on docs/ref/contrib/syndication.txt:
```diff
@@ -1095,3 +1112,47 @@ For example, you might start implementing an iTunes RSS feed generator like so::
+
+.. _xslt stylesheet: https://developer.mozilla.org/en-US/docs/Web/XSLT/Transforming_XML_with_XSLT
+
+.. versionadded:: 5.1
+
+The ability to specify ``xsl_stylesheet_url`` was added.
```
This should go to the top:
```
Feed stylesheets
----------------

.. versionadded:: 5.1

```

yuvadm commented:
Nice to see this request popping up again after 14 years :joy: https://code.djangoproject.com/ticket/12978

kezabelle commented on docs/ref/contrib/syndication.txt:
```diff
@@ -596,6 +596,22 @@ This example illustrates all possible attributes and methods for a
+        def xsl_stylesheet_url(self, obj):
+            """
+            Takes the object returned by get_object() and returns the feed's
+            stylesheet URL (only XSl is supported)
```
```suggestion
            stylesheet URL (only XSL is supported)
```

kezabelle commented on docs/ref/contrib/syndication.txt:
```diff
@@ -1095,3 +1112,47 @@ For example, you might start implementing an iTunes RSS feed generator like so::
+Feed stylesheets
+----------------
+
+You can add styling to your RSS feed using an `XSLT stylesheet`_ by passing
+an ``xsl_stylesheet_url`` argument to the feed's constructor.
```
Just highlighting this, it may well be fine -- `CTRL-F` suggests this is the only place in the document where we refer to the `__init__` as a _constructor_, I can't tell at a glance what the rest of the document uses as nomenclature 🤷

bmispelon replied:
Thanks for pointing this out, I had actually gotten myself mixed up between `django.contrib.syndication.views.Feed` and `django.utils.feedgenerator.SyndicationFeed` and the documentation I'd written didn't actually work.

`SyndicationFeed` is the one that takes arguments in its constructor (this behavior worked and was tested) but `Feed` is customized by settings arguments on the class (this behavior had not been tested and did not work).

I've fixed the missing code and updated the docs 👍🏻

kezabelle commented on docs/ref/contrib/syndication.txt:
```diff
@@ -1095,3 +1112,47 @@ For example, you might start implementing an iTunes RSS feed generator like so::
+Another option is to have a view in your project that renders the XSLT document.
+You can then link it like so::
+
+    from django.contrib.syndication.views import Feed
+    from django.urls import reverse
+
+
+    class FeedWithStylesheetView(Feed):
+        ...  # author, etc.
+        xsl_stylesheet_url = reverse("your-custom-view-name")
```
🤔  Does this example need to use `reverse_lazy` by convention, or is continuing to use `reverse`[^1] going to be OK here?

[^1]: As earlier in the document, but at method evaluation level, not at module level.

kezabelle commented on django/utils/feedgenerator.py:
```diff
@@ -166,6 +168,12 @@ def add_root_elements(self, handler):
         pass
 
+    def add_stylesheets(self, handler):
+        """
+        Add stylesheet(s) to the feed (XSLT).
+        """
```
The corresponding doc you've added in `docs/ref/utils.txt` says: 
> [...] Called from ``write()``

I would assume then, that like the methods just above this, the docstring itself ought to say the same for consistency? Maybe? 🤷

bmispelon commented:
> Nice to see this request popping up again after 14 years 😂 https://code.djangoproject.com/ticket/12978

Oh, I didn't realize there was already a ticket for this (I did search before creating one but clearly not well enough).
Seems like my approach is quite similar to the one you suggested 14 years ago, except I only added support for a single stylesheet, and only for the XSL format.

Do you think those are reasonable limitations @yuvadm ? Your proposal was for multiple RSS stylesheets but from the little I've seen, XSL seems like a more widespread format, do you agree?

yuvadm commented:
@bmispelon after so much time, I hardly remember why I added support for multiple stylesheets, I was probably aiming for maximum flexibility.

In terms of API design I'd probably be liberal in accepting both single and iterable stylesheets in one arg, but unsure if that meets Django code standards.

Anyway nice to see great minds thinking alike :smile:

bmispelon commented:
> Hey @bmispelon, I don't mean to rush you but wanted to mention that when this is ready for a review, please update the ticket flags in trac so it shows in the review queue. Thanks!

Thanks for the heads up, I hadn't notice that the original ticket still had the "patch needs improvement" flag. I've squashed and rebased this PR and updated the flags on the ticket, it should be ready to review.

adamchainz reviewed (commented):
Very thorough PR, looks generally good to me, although I'm not familiar with XSLT stylesheets.

sarahboyce commented on docs/releases/5.1.txt:
```diff
@@ -158,6 +158,10 @@ Minor features
+* All :class:`~django.utils.feedgenerator.SyndicationFeed` classes now support
+  an ``xsl_stylesheet_url`` attribute. If specified, an ``<? xml-stylesheet ?>``
+  processing instruction will be added to the top of the document.
+
 * ...
```
```suggestion
```

felixxm commented on docs/ref/contrib/syndication.txt:
```diff
@@ -1095,3 +1112,47 @@ For example, you might start implementing an iTunes RSS feed generator like so::
+        ...  # author, etc.
+        xsl_stylesheet_url = static("rss_styles.xslt")
+
+
```
```suggestion
```

felixxm commented on docs/ref/utils.txt:
```diff
@@ -368,6 +372,11 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
         Called from ``write()``.
 
+    .. method:: add_stylesheets(self, handler)
+
```
```suggestion

.. versionadded:: 5.1

```

felixxm commented on docs/ref/utils.txt:
```diff
@@ -350,6 +350,10 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
 
+        .. versionadded:: 5.1
+
+        The ``xsl_stylesheet_url`` argument was added.
```
```suggestion
            The ``xsl_stylesheet_url`` argument was added.
```

felixxm commented on docs/ref/contrib/syndication.txt:
```diff
@@ -1095,3 +1112,47 @@ For example, you might start implementing an iTunes RSS feed generator like so::
+
+.. versionadded:: 5.1
+
+    The ability to specify ``xsl_stylesheet_url`` was added.
+
```
This is unnecessary
```suggestion
```

sarahboyce commented:
> Do you think those are reasonable limitations @yuvadm ? Your proposal was for multiple RSS stylesheets but from the little I've seen, XSL seems like a more widespread format, do you agree?

I think a css stylesheet should also be supported 🤔 
- https://jonchristopher.us/blog/beginning-to-style-your-rss-feed/
- https://www.petefreitag.com/blog/css-stylesheet-rss/

bmispelon commented:
I'm a bit puzzled about the test failure. The failing test passes for me locally 🤔

sarahboyce commented on django/utils/feedgenerator.py:
```diff
@@ -57,6 +58,40 @@ def get_tag_uri(url, date):
+        self.url = iri_to_uri(url)
+
+        if mimetype is NOT_PROVIDED:
+            mimetype, _ = mimetypes.guess_type(self.url)
```
This is the main thing that causes the break in the windows tests I think
mimetype is "guessed" as xml on windows

sarahboyce replied:
:very_happy_face: nice work @bmispelon - tricky one! 🥳

sarahboyce commented on tests/syndication_tests/tests.py:
```diff
@@ -561,6 +564,85 @@ def test_feed_no_content_self_closing_tag(self):
+                stylesheet=stylesheet,
+            )
+            doc = feed.writeString("utf-8")
+            with self.subTest(stylesheet=stylesheet):
```
```suggestion
            with self.subTest(expected=expected):
```
The windows test throw the reverse url not found error due to the string method of `Stylesheet`, change to this and you get a more useful error.
Might be worth adding tests specifically for `__str__` and `__repr__` for `Stylesheet`

sarahboyce commented on django/contrib/syndication/views.py:
```diff
@@ -160,6 +160,7 @@ def get_feed(self, obj, request):
             feed_copyright=self._get_dynamic_attr("feed_copyright", obj),
             feed_guid=self._get_dynamic_attr("feed_guid", obj),
             ttl=self._get_dynamic_attr("ttl", obj),
+            stylesheet=self._get_dynamic_attr("stylesheet", obj),
```
Do we want to support having multiple stylesheets?

bmispelon replied:
I wanted to keep the API simple because I assume most people only really have a single stylesheet. I pushed a separate commit that adds documentation (and tests) for how to extend the `Feed` class to support multiple stylesheets.

sarahboyce replied:
Thank you 👍 that's a lovely addition
This is what I'm thinking. 
If we think we might support multiple in the future, we'd want to deprecate `stylesheet` and add `stylesheets`. In that case it's easier to do now.
Otherwise, as you have documented how to do this in your own code, we'd be recommending users that it should be in your own code and that's a decision to not add it into Django.

bmispelon replied:
I've added support for multiple stylesheets (I tried to contain it to a single commit so you can see the diff easier), it didn't add much complexity and your argument about future-proofing against a possible deprecation cycle is convincing.

I did opt to support both a single stylesheet and a list of them, which does make the code (and the documentation) a bit more complex, but I thought it was worth it so that we can offer a simpler API when people only want to use a single stylesheet. If you think that tradeoff is not worth it, I'm also happy to make the API list-only.

sarahboyce replied:
I don't have too strong of an opinion here. I personally would have just supported lists/tuples (a bit like admin inlines) but I don't mind having this. We can get another opinion 👍

yuvadm replied:
With the name `stylesheets` I'd expect to pass an iterable, as a user my immediate reaction is to do `stylesheets=["style.css"]`

moan0s replied:
I'd agree with @yuvadm here :) Thanks so much for working on this PR!

sarahboyce commented on docs/ref/utils.txt:
```diff
@@ -331,6 +331,28 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
+
+.. versionadded:: 5.1
+
+.. class:: Stylesheet
```
I realise this is how it is documented below, but we mostly don't document the `__init__` method this way and do things like:
```suggestion
.. class:: Stylesheet(url, mimetype="", media="screen")
```
And then define each of the arguments like:
```
    .. attribute:: url

        The URL of the stylesheet. This argument is required.
```
I quite like this as you can then link to the specific arguments.
Example: https://docs.djangoproject.com/en/5.0/ref/contrib/postgres/aggregates/#arrayagg

bmispelon replied:
You're right, it's much more usable that way. One thing to note (it might not be a huge deal but I thought I'd flag it just in case): while `media` is indeed a true attribute, the other two are `@property` so if someone ever tries to instantiate a `Stylesheet` then change the attributes afterwords then they'll get an error. It should be possible to add setters to make the attributes work transparently, but I wasn't sure if it was worth the added complexity. What do you think?

sarahboyce replied:
I might be wrong, but I think we do this mostly for formatting rather than to tell the user it is an attribute vs a property 🤔 so I don't think the docs should drive any requirement for setters

bmispelon replied:
Excellent, less work for me 😁

sarahboyce commented on docs/ref/utils.txt:
```diff
@@ -348,7 +370,12 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
+        be a sequence of strings. The ``stylesheet`` argument can be a string
+        or an instance of :class:`Stylesheet`.
+
+        .. versionadded:: 5.1
```
```suggestion
        .. versionchanged:: 5.1
```

bmispelon replied:
Good catch! 👍🏻

sarahboyce commented on docs/ref/contrib/syndication.txt:
```diff
@@ -961,6 +979,7 @@ They share this interface:
+    * ``stylesheet``
 
     Any extra keyword arguments you pass to ``__init__`` will be stored in
     ``self.feed`` for use with `custom feed generators`_.
```
Further down here, we have the same "All parameters should be strings..." note where you might want to add "The ``stylesheet`` argument can be a string or an instance of :class:`Stylesheet`."

bmispelon replied:
Good catch, I've reused the same wording here as in the other file.

bmispelon commented on docs/ref/contrib/syndication.txt:
```diff
@@ -596,6 +596,24 @@ This example illustrates all possible attributes and methods for a
 
         ttl = 600  # Hard-coded Time To Live.
 
+        # STYLESHEET -- One of the following three is optional. The framework
```
Agreed 👍🏻

bmispelon commented on django/utils/feedgenerator.py:
```diff
@@ -247,6 +316,13 @@ def write_items(self, handler):
             handler.endElement("item")
 
+    def start_document(self, handler):
+        handler.startDocument()
```
Good point, this method doesn't really add anything now.

bmispelon commented on django/utils/feedgenerator.py:
```diff
@@ -57,6 +58,56 @@ def get_tag_uri(url, date):
+class Stylesheet:
+    """An RSS stylesheet"""
+
+    def __init__(self, url, mimetype=NOT_PROVIDED, media="screen"):
```
Great idea 👍🏻

sarahboyce commented on docs/releases/5.1.txt:
```diff
@@ -158,7 +158,9 @@ Minor features
 
-* ...
+* All :class:`~django.utils.feedgenerator.SyndicationFeed` classes now support
+  a ``stylesheet`` attribute. If specified, an ``<? xml-stylesheet ?>``
```
```suggestion
  a ``stylesheets`` attribute. If specified, an ``<? xml-stylesheet ?>``
```

bmispelon replied:
Doh, I forgot to update the release notes 🤦🏻 
Thanks!

bmispelon commented on docs/ref/contrib/syndication.txt:
```diff
@@ -596,6 +596,30 @@ This example illustrates all possible attributes and methods for a
+        stylesheets = [
+            feedgenerator.Stylesheet("/stylesheet1.xsl"),
+            feedgenerator.Stylesheet("/stylesheet2.xsl"),
+        ]
```
It was a hard choice, but I went with the 3rd example 😁

bmispelon commented on docs/ref/contrib/syndication.txt:
```diff
@@ -961,14 +985,21 @@ They share this interface:
+    Beware that some control characters are
+    `not allowed <https://www.w3.org/International/questions/qa-controls>`_ in
+    XML documents. If your content has some of them, you might encounter a
     :exc:`ValueError` when producing the feed.
```
Ah yes, good point

bmispelon commented on docs/ref/contrib/syndication.txt:
```diff
@@ -1095,3 +1126,81 @@ For example, you might start implementing an iTunes RSS feed generator like so::
+You can add styling to your RSS feed using a stylesheet (typically in the XSLT_
+or CSS formats) by setting the ``stylesheets`` attribute on the feed class.
+
+This can be a hardcoded URL::
```
I don't think that's necessary. The `::` introduces a code block with the default language (which should be Python). On my local build the code examples are correctly highlighted as Python, and the rest of the document uses the same syntax (except for a few examples which have an explicit `code-block:: django+html`)

bmispelon commented on django/utils/feedgenerator.py:
```diff
@@ -166,6 +223,12 @@ def add_root_elements(self, handler):
+        """
+        Add stylesheet(s) to the feed. Called from write().
+        """
+        pass
```
You're right that the method is overwritten in the `RssFeed` class later in the file.

I copied the style of other existing methods like `add_root_elements()` or `add_item_elements()` which are defined on the base  `SyndicationFeed` class but reimplemented later. I believe this serves as some kind of documentation for what methods should be implemented.

felixxm commented on django/utils/feedgenerator.py:
```diff
@@ -75,12 +123,24 @@ def __init__(
+        if stylesheets is not None:
+            if isinstance(stylesheets, (Stylesheet, str)):
+                raise TypeError(
+                    "stylesheets should be a list, not {stylesheets.__class__}"
```
`f` is missing, so it's also untested
```suggestion
                   f"stylesheets should be a list, not {stylesheets.__class__}"
```

bmispelon replied:
That's what I get for testing the exception and not its message 🤦🏻 

Thanks!

sarahboyce commented on docs/ref/utils.txt:
```diff
@@ -331,6 +331,32 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
+
+.. class:: Stylesheet(url, mimetype="", media="screen")
+
+    Represents an RSS stylesheet
```
```suggestion
    Represents an RSS stylesheet.
```

sarahboyce commented on docs/ref/contrib/syndication.txt:
```diff
@@ -1095,3 +1123,90 @@ For example, you might start implementing an iTunes RSS feed generator like so::
+
+.. versionadded:: 5.1
+
+You can add styling to your RSS feed using a stylesheet (typically in the XSLT_
+or CSS formats) by setting the ``stylesheets`` attribute on the feed class.
```
```suggestion
If you wish to have your RSS feed render nicely in a browser, you will need to
provide styling information for the XML file, typically in XSLT_ or CSS formats.

You can add this to your RSS feed by setting the ``stylesheets`` attribute on the
feed class.
```
Maybe?

sarahboyce reviewed (commented):
This looks great 👍 thank you @bmispelon 
Can you rebase, squash commits, update the version notes to 5.2 and move the release note to 5.2 also? ⭐

bmispelon commented:
> This looks great 👍 thank you @bmispelon Can you rebase, squash commits, update the version notes to 5.2 and move the release note to 5.2 also? ⭐

Sure thing, done 👍🏻

smithdc1 commented on docs/ref/contrib/syndication.txt:
```diff
@@ -596,6 +596,24 @@ This example illustrates all possible attributes and methods for a
+            """
+
+        # Hardcoded stylesheets
+        stylesheets = ["/stylesheet1.xsl", "stylesheet2.xls"]
```
```suggestion
        stylesheets = ["/stylesheet1.xsl", "stylesheet2.xsl"]
```
Typo?

bmispelon replied:
Ah yes, good catch 😅

sarahboyce reviewed (approved):
Thank you @bmispelon for your work on this ⭐ 
This looks good to me 🥳
